log/scope_test.go

```diff
@@ -218,6 +220,130 @@ func TestBasicScopes(t *testing.T) {
 	}
 }
 
+func TestScopeAddLabel(t *testing.T) {
+	const name = "TestScope"
+	const desc = "Desc"
+	s := RegisterScope(name, desc, 0)
+	s.SetOutputLevel(DebugLevel)
+
+	lines, err := captureStdout(func() {
+		funcs.Store(funcs.Load().(patchTable))
+		s.AddLabel("foo", "bar").AddLabel("baz", 123).AddLabel("qux", 0.123).Debuga("Hello")
+		s.Debuga("Hello")
+		_ = Sync()
+	})
+
+	if err != nil {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	expect := `Hello {"foo":"bar","baz":1,"qux":0.123}`
+	mustRegexMatchString(t, lines[0], expect)
+	mustRegexMatchString(t, lines[1], expect)
+}
+
+func TestScopeWithLabel(t *testing.T) {
+	const name = "TestScope"
+	const desc = "Desc"
+	s := RegisterScope(name, desc, 0)
+	s.SetOutputLevel(DebugLevel)
+
+	lines, err := captureStdout(func() {
+		funcs.Store(funcs.Load().(patchTable))
+		s.AddLabel("foo", "bar").WithLabel("baz", 123).Debuga("Hello")
+		s.Debuga("Hello")
+		_ = Sync()
+	})
+	if err != nil {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `Hello {"foo":"bar","baz":123}`)
+	mustRegexMatchString(t, lines[1], `Hello {"foo":"bar"}`)
+}
+
+func TestScopeClearLabel(t *testing.T) {
+	const name = "TestScope"
+	const desc = "Desc"
+	s := RegisterScope(name, desc, 0)
+	s.SetOutputLevel(DebugLevel)
+
+	lines, err := captureStdout(func() {
+		funcs.Store(funcs.Load().(patchTable))
+		s.AddLabel("foo", "bar").AddLabel("baz", 123).AddLabel("qux", 0.123)
+		s.ClearLabelKey("qux")
+		s.Debuga("Hello")
+		s.ClearAllLabels()
+		s.Debuga("Hello")
+
+		_ = Sync()
+	})
+	if err != nil {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `Hello {"foo":"bar","baz":123}`)
+	mustRegexMatchString(t, lines[1], `Hello$`)
+}
+
+func TestScopeLocal(t *testing.T) {
+	const name = "TestScope"
+	const desc = "Desc"
+	s := RegisterScope(name, desc, 0)
+	s.SetOutputLevel(DebugLevel)
+
+	lines, err := captureStdout(func() {
+		funcs.Store(funcs.Load().(patchTable))
+		s.AddLabel("foo", "bar").AddLabel("baz", 123)
+		s2 := s.Local()
+		s2.ClearAllLabels()
+		s.Debuga("Hello")
+		s2.Debuga("Hello")
+
+		_ = Sync()
+	})
+	if err != nil {
+		t.Errorf("Got error '%v', expected success", err)
+	}
+
+	mustRegexMatchString(t, lines[0], `Hello {"foo":"bar","baz":123}`)
```

Comment:
Are there logging ingestion tools that read logs that are partially plaintext and partially json?

Reply:
Yes, but it's up to us to decide exactly how the log outputs. It could be all JSON, text + JSON or all text. That can easily be made more configurable.

Reply:
how come in the previous test its json with spaces and in this one there are no spaces?